Fixes 11053: focus event causing jumpy scroll effect

Andrewgdewar opened the pull request:
Closes #11053 

Prevents inconsistent scrolling behaviour on a number of select/menu components.

This also removes the timeout, which appears is no longer needed. 
(Slightly snappier without, with/without barely noticeable).

**What**: Closes 11053

**Additional issues**:
[Associated External issue from Red Hat Insights team](https://issues.redhat.com/browse/HMS-4746)

patternfly-build commented:
Preview: https://patternfly-react-pr-11056.surge.sh

Andrewgdewar commented:
Someone please point me to the documentation to run the linter.
🙄

thatblindgeye reviewed (changes requested):
Some initial comments below that would most likely apply to all components updated.

For the linter docs, should be the last bullet in the [React component requirements](https://github.com/patternfly/patternfly-react/blob/main/CONTRIBUTING.md#react-component-requirements) of our contributing guide.

Andrewgdewar commented on packages/react-core/src/components/Dropdown/Dropdown.tsx:
```diff
@@ -114,20 +114,18 @@ const DropdownBase: React.FunctionComponent<DropdownProps> = ({
         if (onOpenChangeKeys.includes(event.key)) {
           onOpenChange(false);
-          toggleRef.current?.focus();
+          toggleRef.current?.focus({preventScroll:true});
```
I was still seeing the described problem on menudropdowns on filter views without the above change. 
Why this is being triggered on multi-select components I don't know. 
With the other fix in place, and this one not. 
The same behaviour below was occurring:

https://github.com/user-attachments/assets/f29b5fde-5d15-4ae8-8887-e322a14d0af0

Andrewgdewar replied:
[This](https://github.com/content-services/content-sources-frontend/blob/a88fae033f857668c472a7cfe3e707f2a809b05e/src/Pages/Repositories/ContentListTable/components/SnapshotDetailsModal/Tabs/SnapshotErrataFilters.tsx#L166) is the above component.

[This](https://github.com/content-services/content-sources-frontend/blob/main/src/components/DropdownSelect/DropdownSelect.tsx#L28) is the wrapper I've built to contain the madness that is the patternfly Select component. 😆

thatblindgeye replied:
So I tried mimicking the setup from both files linked above just within a PatternFly example (wasn't able to get the repo linked in the above files running locally), and I'm not able to reproduce the jump scroll when simply removing the `preventScroll` option from this toggleRef.focus call of the Select component. May not be an exact 1:1 mimic, and it could also be a difference between the PatternFly workspace and yours, but as far as I can tell reverting this specific line in the changed files shouldn't cause that jump scroll bug when the menu is opened.

Personally I would opt to remove the options being passed to the `toggleRef.focus()` calls in these files (but keep the option in the other focus calls, i.e. `firstElement...focus()`), as the expected behavior should be that the page will scroll back to where the MenuToggle receives focus upon the Menu being closed via keyboard. Then if the jump scroll bug is still occurring with the rest of these updates merged in, we could revisit again to investigate further or add the options in as you did originally here.

How did you go about testing the updates in this PR in your product running locally? And you can confirm that reverting the `toggleRef.current?.focus({preventScroll:true});` lines in this PR back to just `toggleRef.current?.focus();` (but keeping all other changes) causes the bug to still occur when clicking on a MenuToggle to open the select/dropdown/menu?

Andrewgdewar replied:
>>How did you go about testing the updates 
The other component changes were not tested; the copied logic matched and thus I added the changes as I presumed they would be desired. 

>>How did you go about testing the updates in this PR
I simply made the changes to the patternfly components in node_modules > patternfly, and rebuilt. 

>> Personally I would opt to remove the options being passed to the toggleRef.focus()
This is fine, I'll either use another component or find another way to prevent the behaviour.

>>can confirm that reverting the toggleRef.current?.focus({preventScroll:true})
I'll test again today

thatblindgeye replied:
Hey @Andrewgdewar , just wanted to check in on if you were able to test the above?

Also wanted to clarify in case I might be misinterpreting the issue here. Can you confirm between the foilowing which you originally meant in regards to this `toggleRef...focus` call needing the preventScroll option?

1. When the toggleRef focus call does NOT have the preventScroll option passed in, the jump scroll bug still occurs when the menu/select/drodown is opened
2. Or, were you saying that you believe that we shouldn't scroll the page when the menu/select/dropdown is closed and the MenuToggle (toggleRef in this case) receives focus?

Andrewgdewar replied:
> When the toggleRef focus call does NOT have the preventScroll option passed in, the jump scroll bug still occurs when the menu/select/drodown is opened

Andrewgdewar commented on packages/react-core/src/components/Dropdown/Dropdown.tsx:
```diff
@@ -114,20 +114,18 @@ const DropdownBase: React.FunctionComponent<DropdownProps> = ({
     const handleClick = (event: MouseEvent) => {
       // toggle was opened, focus on first menu item
       if (isOpen && shouldFocusFirstItemOnOpen && toggleRef.current?.contains(event.target as Node)) {
-        setTimeout(() => {
```
What problem was this attempting to solve, as I was able to use the component without this, with an apparent performance boost (of 10ms). 
In the case highlighted, using loading state, I'm unsure how the 10ms wait time could possibly assist in focusing, as the dependency on the API would likely be much longer than that.

thatblindgeye replied:
Originally the 10ms timeout was attempting to resolve the scroll jump bug, which was confirmed to have worked in at least 1 product. Outside of that, keeping the timeout for a load state might be more advantageous if we implement a prop to control the timeout rather than hardcoding it to 10ms (currently tracked by issue https://github.com/patternfly/patternfly-react/issues/11020).

@tlabaj would we still want to implement a prop for controlling the timeout duration, or do we want to completely remove the setTimeout and close out 11020 (if a consumer wants a load state sort of implementation, do we just point to our custom menu examples that would require them to wire something up)? Maybe depends on whether we can get confirmation from other products that this new approach resolves the issue for them.

Andrewgdewar replied:
I can re-add the timeout with a prop for the timing.

Andrewgdewar replied:
Done!

Andrewgdewar commented on packages/react-core/src/components/Dropdown/Dropdown.tsx:
```diff
@@ -114,20 +114,18 @@ const DropdownBase: React.FunctionComponent<DropdownProps> = ({
           );
-          firstElement && (firstElement as HTMLElement).focus();
-        }, 10);
+          firstElement && (firstElement as HTMLElement).focus({preventScroll:true});
```
The only issue I have with that, is the scrolling is super inconsistent with setTimeout on the component. 
It leads to the user expecting there to be scroll events if preventScrollOnItemFocus = false, when the behaviour is inconsistent.

Andrewgdewar commented:
Updated:
- Lint
- Added variable `preventScrollOnItemFocus` (feedback on copy welcome)
- setTimeout is removed, waiting on explanation of it's requirement, setting `preventScrollOnItemFocus` to false with setTimeout in place creates inconsistent scroll behaviour.

Andrewgdewar commented:
/retest

tlabaj commented:
@Andrewgdewar try and rebase your PR to fix the test failure. I pushed a change yesterday.

tlabaj commented on packages/react-core/src/components/Menu/MenuContainer.tsx:
```diff
@@ -37,6 +37,8 @@ export interface MenuContainerProps {
   /** Additional properties to pass to the Popper */
   popperProps?: MenuPopperProps;
+  /** Flag indicating if scroll on focus of the first menu item should occur. */
+  preventScrollOnItemFocus?: boolean;
```
nit. Same comment here about the name of this prop. Can we prefix it with `should`?

tlabaj commented on packages/react-core/src/components/Pagination/PaginationOptionsMenu.tsx:
```diff
@@ -56,6 +56,8 @@ export interface PaginationOptionsMenuProps extends React.HTMLProps<HTMLDivEleme
   /** @beta The container to append the pagination options menu to. Overrides the containerRef prop. */
   appendTo?: HTMLElement | (() => HTMLElement) | 'inline';
+  /** Flag indicating if scroll on focus of the first menu item should occur. */
+  preventScrollOnItemFocus?: boolean;
```
nit. Same comment here about the name of this prop. Can we prefix it with `should`?

tlabaj commented on packages/react-core/src/components/Tabs/OverflowTab.tsx:
```diff
@@ -21,6 +21,8 @@ export interface OverflowTabProps extends React.HTMLProps<HTMLLIElement> {
   /** z-index of the overflow tab */
   zIndex?: number;
+  /** Flag indicating if scroll on focus of the first menu item should occur. */
+  preventScrollOnItemFocus?: boolean;
```
nit. Same comment here about the name of this prop. Can we prefix it with `should`?

tlabaj commented on packages/react-core/src/components/Select/Select.tsx:
```diff
@@ -78,6 +78,8 @@ export interface SelectProps extends MenuProps, OUIAProps {
   /** Indicates if the select menu should be scrollable */
   isScrollable?: boolean;
+  /** Flag indicating if scroll on focus of the first menu item should occur. */
+  preventScrollOnItemFocus?: boolean;
```
nit. Same comment here about the name of this prop. Can we prefix it with `should`?

tlabaj reviewed (changes requested):
I few nits

Andrewgdewar commented:
> I few nits

Thanks for the review @tlabaj!

thatblindgeye commented on packages/react-core/src/components/Select/Select.tsx:
```diff
@@ -78,6 +78,8 @@ export interface SelectProps extends MenuProps, OUIAProps {
   /** Indicates if the select menu should be scrollable */
   isScrollable?: boolean;
+  /** Flag indicating if scroll on focus of the first menu item should occur. */
+  shouldpreventScrollOnItemFocus?: boolean;
```
Small casing issue in all of the changed files:

```suggestion
  shouldPreventScrollOnItemFocus?: boolean;
```

Andrewgdewar replied:
Oh silliness! Derp!

thatblindgeye commented on packages/react-core/src/components/Tabs/OverflowTab.tsx:
```diff
@@ -21,6 +21,8 @@ export interface OverflowTabProps extends React.HTMLProps<HTMLLIElement> {
   /** z-index of the overflow tab */
   zIndex?: number;
+  /** Flag indicating if scroll on focus of the first menu item should occur. */
+  shouldpreventScrollOnItemFocus?: boolean;
```
```suggestion
  shouldPreventScrollOnItemFocus?: boolean;
```

thatblindgeye commented on packages/react-core/src/components/Menu/MenuContainer.tsx:
```diff
@@ -37,6 +37,8 @@ export interface MenuContainerProps {
   /** Additional properties to pass to the Popper */
   popperProps?: MenuPopperProps;
+  /** Flag indicating if scroll on focus of the first menu item should occur. */
+  shouldpreventScrollOnItemFocus?: boolean;
```
```suggestion
  shouldPreventScrollOnItemFocus?: boolean;
```

thatblindgeye commented on packages/react-core/src/components/Pagination/PaginationOptionsMenu.tsx:
```diff
@@ -56,6 +56,8 @@ export interface PaginationOptionsMenuProps extends React.HTMLProps<HTMLDivEleme
   /** @beta The container to append the pagination options menu to. Overrides the containerRef prop. */
   appendTo?: HTMLElement | (() => HTMLElement) | 'inline';
+  /** Flag indicating if scroll on focus of the first menu item should occur. */
+  shouldpreventScrollOnItemFocus?: boolean;
```
```suggestion
  shouldPreventScrollOnItemFocus?: boolean;
```

thatblindgeye commented on packages/react-core/src/components/Dropdown/Dropdown.tsx:
```diff
@@ -71,6 +71,8 @@ export interface DropdownProps extends MenuProps, OUIAProps {
   /** @beta Flag indicating the first menu item should be focused after opening the dropdown. */
   shouldFocusFirstItemOnOpen?: boolean;
+  /** Flag indicating if scroll on focus of the first menu item should occur. */
+  shouldpreventScrollOnItemFocus?: boolean;
```
```suggestion
  shouldPreventScrollOnItemFocus?: boolean;
```

thatblindgeye reviewed (changes requested):
Small casing fix noted below (usage of the prop will also need to be updated), and also left additional comments on previous replies above.

Andrewgdewar commented:
> Small casing fix noted below (usage of the prop will also need to be updated), and also left additional comments on previous replies above.

Thanks! I took the recommended variable too literally (lol.. copied casing).
I also re-added the timeout variable as requested, let me know if copy is okay for that variable name/description.

patternfly-build commented:
Your changes have been released in: 
 - @patternfly/react-code-editor@5.4.9
 - @patternfly/react-core@5.4.7
 - @patternfly/react-docs@6.4.11
 - @patternfly/react-drag-drop@5.4.7
 - demo-app-ts@5.4.10
 - @patternfly/react-table@5.4.7
 - @patternfly/react-templates@1.1.7

Thanks for your contribution! :tada:
